### ISSUE TITLE
AdminClient is not garbage-collected after going out-of-scope

### ISSUE THREAD

**Issue by ckataki** [confluentinc/confluent-kafka-python#975]
Description
===========
The admin client keeps running in the background even after going out of scope. The following function is a simplified version of our internal code. The `tester()` function creates a locally-scoped AdminClient, and expects it to be destroyed when it exits.

In practice, we see that the client is not GC'ed (seems to stay in generation 0), and keeps polling the brokers. We have verified this by killing the brokers; this leads to a whole bunch of connection errors in the librdkafka logs, long after we expect the client to have been GC'ed.
 If we manually set the local variable to None (`admin_client = None`), things works as expected.

Are we misunderstanding something here, or mis-using the library? We think there's a similar issue with the Producer as well, but have not done any testing to verify. The Consumer works okay (after calling the `close()` method).

How to reproduce
================
```
import gc
import time

from confluent_kafka import KafkaError, KafkaException, version, libversion
from confluent_kafka.admin import AdminClient, NewTopic


def tester():
    print("Inside tester() function. Admin client will be created here.")
    admin_client = AdminClient({
        "bootstrap.servers": "kafka-1:19092,kafka-2:29092,kafka-3:39092",
        "client.id": "Tester",
        "debug": "broker",
    })
    topic_config = {
        "topic": "tester-topic",
        "num_partitions": 3,
        "replication_factor": 1,
        "config": {
            "min.insync.replicas": 1,
            "message.timestamp.type": "LogAppendTime",
        },
    }
    topic = NewTopic(**topic_config)
    try:
        futures = admin_client.create_topics([topic])
    except Exception as exc:
        print(f"Exception raised creating topics: {exc}")
        return
    try:
        futures[topic_config["topic"]].result()
        print("Future result returned")
    except KafkaException as ex:
        if ex.args[0].code() != KafkaError.TOPIC_ALREADY_EXISTS:
            print(f"{ex} while creating topic")
        else:
            print("Topic already exists.")
    except Exception as ex:
        print(f"Exception {ex}")

    while admin_client.poll(1):
        print("Defensive polling...")

    # Uncommenting this line "fixes" the issue
    # admin_client = None
    print("Last line of tester() function. Admin client should go out of scope here.")

if __name__ == "__main__":
    print(f"Version: {version()}, Lib version: {libversion()}")
    tester()
    print("Tester function has returned. All variables should be GC'ed by now.")
    try:
        while True:
            print("AdminClient in GC-0", "AdminClient" in [obj.__class__.__name__ for obj in gc.get_objects(0)])
            print("AdminClient in GC-1", "AdminClient" in [obj.__class__.__name__ for obj in gc.get_objects(1)])
            print("AdminClient in GC-2", "AdminClient" in [obj.__class__.__name__ for obj in gc.get_objects(2)])
            time.sleep(5)
    except KeyboardInterrupt:
        print("Done")
```

Client logs without explicitly setting to None
=================================
```
Version: ('1.5.0', 17104896), Lib version: ('1.5.0', 17105151)
Inside tester() function. Admin client will be created here.
%7|1603848614.742|BROKER|Tester#producer-1| [thrd:app]: kafka-1:19092/bootstrap: Added new broker with NodeId -1
%7|1603848614.742|BROKER|Tester#producer-1| [thrd:app]: kafka-2:29092/bootstrap: Added new broker with NodeId -1
%7|1603848614.742|BRKMAIN|Tester#producer-1| [thrd::0/internal]: :0/internal: Enter main broker thread
%7|1603848614.743|BRKMAIN|Tester#producer-1| [thrd:kafka-1:19092/bootstrap]: kafka-1:19092/bootstrap: Enter main broker thread
%7|1603848614.743|BROKER|Tester#producer-1| [thrd:app]: kafka-3:39092/bootstrap: Added new broker with NodeId -1
%7|1603848614.744|CONNECT|Tester#producer-1| [thrd:app]: kafka-3:39092/bootstrap: Selected for cluster connection: bootstrap servers added (broker has 0 connection attempt(s))
%7|1603848614.743|BRKMAIN|Tester#producer-1| [thrd:kafka-2:29092/bootstrap]: kafka-2:29092/bootstrap: Enter main broker thread
%7|1603848614.744|BRKMAIN|Tester#producer-1| [thrd:kafka-3:39092/bootstrap]: kafka-3:39092/bootstrap: Enter main broker thread
%7|1603848614.744|CONNECT|Tester#producer-1| [thrd:kafka-3:39092/bootstrap]: kafka-3:39092/bootstrap: Received CONNECT op
%7|1603848614.744|INIT|Tester#producer-1| [thrd:app]: librdkafka v1.5.0 (0x10500ff) Tester#producer-1 initialized (builtin.features gzip,snappy,ssl,sasl,regex,lz4,sasl_plain,sasl_scram,plugins,zstd,sasl_oauthbearer, STATIC_LINKING GCC GXX PKGCONFIG INSTALL GNULD LDS LIBDL PLUGINS STATIC_LIB_zlib ZLIB STATIC_LIB_libcrypto STATIC_LIB_libssl SSL STATIC_LIB_libzstd ZSTD HDRHISTOGRAM SYSLOG SNAPPY SOCKEM SASL_SCRAM SASL_OAUTHBEARER CRC32C_HW, debug 0x2)
%7|1603848614.745|CONNECT|Tester#producer-1| [thrd:main]: Not selecting any broker for cluster connection: still suppressed for 49ms: lookup controller
%7|1603848614.745|STATE|Tester#producer-1| [thrd:kafka-3:39092/bootstrap]: kafka-3:39092/bootstrap: Broker changed state INIT -> TRY_CONNECT
%7|1603848614.745|CONNECT|Tester#producer-1| [thrd:kafka-3:39092/bootstrap]: kafka-3:39092/bootstrap: broker in state TRY_CONNECT connecting
%7|1603848614.745|STATE|Tester#producer-1| [thrd:kafka-3:39092/bootstrap]: kafka-3:39092/bootstrap: Broker changed state TRY_CONNECT -> CONNECT
%7|1603848614.745|CONNECT|Tester#producer-1| [thrd:main]: Not selecting any broker for cluster connection: still suppressed for 48ms: lookup controller
%7|1603848614.745|CONNECT|Tester#producer-1| [thrd:main]: Not selecting any broker for cluster connection: still suppressed for 48ms: lookup controller
%7|1603848614.747|CONNECT|Tester#producer-1| [thrd:kafka-3:39092/bootstrap]: kafka-3:39092/bootstrap: Connecting to ipv4#172.23.0.9:39092 (plaintext) with socket 11
%7|1603848614.749|CONNECT|Tester#producer-1| [thrd:kafka-3:39092/bootstrap]: kafka-3:39092/bootstrap: Connected to ipv4#172.23.0.9:39092
%7|1603848614.749|CONNECTED|Tester#producer-1| [thrd:kafka-3:39092/bootstrap]: kafka-3:39092/bootstrap: Connected (#1)
%7|1603848614.750|FEATURE|Tester#producer-1| [thrd:kafka-3:39092/bootstrap]: kafka-3:39092/bootstrap: Updated enabled protocol features +ApiVersion to ApiVersion
%7|1603848614.750|STATE|Tester#producer-1| [thrd:kafka-3:39092/bootstrap]: kafka-3:39092/bootstrap: Broker changed state CONNECT -> APIVERSION_QUERY
%7|1603848614.751|CONNECT|Tester#producer-1| [thrd:main]: Not selecting any broker for cluster connection: still suppressed for 43ms: lookup controller
%7|1603848614.752|FEATURE|Tester#producer-1| [thrd:kafka-3:39092/bootstrap]: kafka-3:39092/bootstrap: Updated enabled protocol features to MsgVer1,ApiVersion,BrokerBalancedConsumer,ThrottleTime,Sasl,SaslHandshake,BrokerGroupCoordinator,LZ4,OffsetTime,MsgVer2,IdempotentProducer,ZSTD,SaslAuthReq
%7|1603848614.752|STATE|Tester#producer-1| [thrd:kafka-3:39092/bootstrap]: kafka-3:39092/bootstrap: Broker changed state APIVERSION_QUERY -> UP
%7|1603848614.753|CLUSTERID|Tester#producer-1| [thrd:main]: kafka-3:39092/bootstrap: ClusterId update "" -> "JA1lgI6fTvWwm5lTvT7ETA"
%7|1603848614.754|CONTROLLERID|Tester#producer-1| [thrd:main]: kafka-3:39092/bootstrap: ControllerId update -1 -> 2
%7|1603848614.755|UPDATE|Tester#producer-1| [thrd:kafka-2:29092/bootstrap]: kafka-2:29092/bootstrap: NodeId changed from -1 to 2
%7|1603848614.755|UPDATE|Tester#producer-1| [thrd:kafka-2:29092/bootstrap]: kafka-2:29092/2: Name changed from kafka-2:29092/bootstrap to kafka-2:29092/2
%7|1603848614.755|LEADER|Tester#producer-1| [thrd:kafka-2:29092/bootstrap]: kafka-2:29092/2: Mapped 0 partition(s) to broker
%7|1603848614.755|CONNECT|Tester#producer-1| [thrd:kafka-2:29092/bootstrap]: kafka-2:29092/2: Received CONNECT op
%7|1603848614.756|STATE|Tester#producer-1| [thrd:kafka-2:29092/bootstrap]: kafka-2:29092/2: Broker changed state INIT -> TRY_CONNECT
%7|1603848614.756|CONNECT|Tester#producer-1| [thrd:kafka-2:29092/bootstrap]: kafka-2:29092/2: broker in state TRY_CONNECT connecting
%7|1603848614.756|STATE|Tester#producer-1| [thrd:kafka-2:29092/bootstrap]: kafka-2:29092/2: Broker changed state TRY_CONNECT -> CONNECT
%7|1603848614.756|UPDATE|Tester#producer-1| [thrd:kafka-3:39092/bootstrap]: kafka-3:39092/bootstrap: NodeId changed from -1 to 3
%7|1603848614.756|UPDATE|Tester#producer-1| [thrd:kafka-3:39092/bootstrap]: kafka-3:39092/3: Name changed from kafka-3:39092/bootstrap to kafka-3:39092/3
%7|1603848614.756|LEADER|Tester#producer-1| [thrd:kafka-3:39092/bootstrap]: kafka-3:39092/3: Mapped 0 partition(s) to broker
%7|1603848614.756|STATE|Tester#producer-1| [thrd:kafka-3:39092/bootstrap]: kafka-3:39092/3: Broker changed state UP -> UPDATE
%7|1603848614.756|STATE|Tester#producer-1| [thrd:kafka-3:39092/bootstrap]: kafka-3:39092/3: Broker changed state UPDATE -> UP
%7|1603848614.757|UPDATE|Tester#producer-1| [thrd:kafka-1:19092/bootstrap]: kafka-1:19092/bootstrap: NodeId changed from -1 to 1
%7|1603848614.757|UPDATE|Tester#producer-1| [thrd:kafka-1:19092/bootstrap]: kafka-1:19092/1: Name changed from kafka-1:19092/bootstrap to kafka-1:19092/1
%7|1603848614.758|LEADER|Tester#producer-1| [thrd:kafka-1:19092/bootstrap]: kafka-1:19092/1: Mapped 0 partition(s) to broker
%7|1603848614.757|CONNECT|Tester#producer-1| [thrd:kafka-2:29092/bootstrap]: kafka-2:29092/2: Connecting to ipv4#172.23.0.7:29092 (plaintext) with socket 12
%7|1603848614.759|CONNECT|Tester#producer-1| [thrd:kafka-2:29092/bootstrap]: kafka-2:29092/2: Connected to ipv4#172.23.0.7:29092
%7|1603848614.759|CONNECTED|Tester#producer-1| [thrd:kafka-2:29092/bootstrap]: kafka-2:29092/2: Connected (#1)
%7|1603848614.759|FEATURE|Tester#producer-1| [thrd:kafka-2:29092/bootstrap]: kafka-2:29092/2: Updated enabled protocol features +ApiVersion to ApiVersion
%7|1603848614.759|STATE|Tester#producer-1| [thrd:kafka-2:29092/bootstrap]: kafka-2:29092/2: Broker changed state CONNECT -> APIVERSION_QUERY
%7|1603848614.760|FEATURE|Tester#producer-1| [thrd:kafka-2:29092/bootstrap]: kafka-2:29092/2: Updated enabled protocol features to MsgVer1,ApiVersion,BrokerBalancedConsumer,ThrottleTime,Sasl,SaslHandshake,BrokerGroupCoordinator,LZ4,OffsetTime,MsgVer2,IdempotentProducer,ZSTD,SaslAuthReq
%7|1603848614.761|STATE|Tester#producer-1| [thrd:kafka-2:29092/bootstrap]: kafka-2:29092/2: Broker changed state APIVERSION_QUERY -> UP
Topic already exists.
Last line of tester() function. Admin client should go out of scope here.
Tester function has returned. All variables should be GC'ed by now.
AdminClient in GC-0 True
AdminClient in GC-1 False
AdminClient in GC-2 False
AdminClient in GC-0 True
AdminClient in GC-1 False
AdminClient in GC-2 False
^CDone
%7|1603848621.458|DESTROY|Tester#producer-1| [thrd:app]: Terminating instance (destroy flags none (0x0))
%7|1603848621.458|DESTROY|Tester#producer-1| [thrd:main]: Destroy internal
%7|1603848621.458|DESTROY|Tester#producer-1| [thrd:main]: Waiting for background queue thread to terminate
%7|1603848621.458|DESTROY|Tester#producer-1| [thrd:main]: Removing all topics
%7|1603848621.459|DESTROY|Tester#producer-1| [thrd:main]: Sending TERMINATE to kafka-3:39092/3
%7|1603848621.459|TERM|Tester#producer-1| [thrd:kafka-3:39092/bootstrap]: kafka-3:39092/3: Received TERMINATE op in state UP: 1 refcnts, 0 toppar(s), 0 active toppar(s), 0 outbufs, 0 waitresps, 0 retrybufs
%7|1603848621.462|FAIL|Tester#producer-1| [thrd:kafka-3:39092/bootstrap]: kafka-3:39092/3: Client is terminating (after 6705ms in state UP) (_DESTROY)
%7|1603848621.462|STATE|Tester#producer-1| [thrd:kafka-3:39092/bootstrap]: kafka-3:39092/3: Broker changed state UP -> DOWN
%7|1603848621.462|TERMINATE|Tester#producer-1| [thrd:kafka-3:39092/bootstrap]: kafka-3:39092/3: Handle is terminating in state DOWN: 1 refcnts (0x563f68e1a410), 0 toppar(s), 0 active toppar(s), 0 outbufs, 0 waitresps, 0 retrybufs: failed 0 request(s) in retry+outbuf
%7|1603848621.462|FAIL|Tester#producer-1| [thrd:kafka-3:39092/bootstrap]: kafka-3:39092/3: Broker handle is terminating (after 0ms in state DOWN) (_DESTROY)
%7|1603848621.459|DESTROY|Tester#producer-1| [thrd:main]: Sending TERMINATE to kafka-2:29092/2
%7|1603848621.463|DESTROY|Tester#producer-1| [thrd:main]: Sending TERMINATE to kafka-1:19092/1
%7|1603848621.464|TERM|Tester#producer-1| [thrd:kafka-1:19092/bootstrap]: kafka-1:19092/1: Received TERMINATE op in state INIT: 1 refcnts, 0 toppar(s), 0 active toppar(s), 0 outbufs, 0 waitresps, 0 retrybufs
%7|1603848621.464|FAIL|Tester#producer-1| [thrd:kafka-1:19092/bootstrap]: kafka-1:19092/1: Client is terminating (after 6722ms in state INIT) (_DESTROY)
%7|1603848621.464|STATE|Tester#producer-1| [thrd:kafka-1:19092/bootstrap]: kafka-1:19092/1: Broker changed state INIT -> DOWN
%7|1603848621.465|TERMINATE|Tester#producer-1| [thrd:kafka-1:19092/bootstrap]: kafka-1:19092/1: Handle is terminating in state DOWN: 1 refcnts (0x563f68e18970), 0 toppar(s), 0 active toppar(s), 0 outbufs, 0 waitresps, 0 retrybufs: failed 0 request(s) in retry+outbuf
%7|1603848621.465|FAIL|Tester#producer-1| [thrd:kafka-1:19092/bootstrap]: kafka-1:19092/1: Broker handle is terminating (after 0ms in state DOWN) (_DESTROY)
%7|1603848621.466|TERM|Tester#producer-1| [thrd::0/internal]: :0/internal: Received TERMINATE op in state INIT: 1 refcnts, 0 toppar(s), 0 active toppar(s), 0 outbufs, 0 waitresps, 0 retrybufs
%7|1603848621.466|FAIL|Tester#producer-1| [thrd::0/internal]: :0/internal: Client is terminating (after 6725ms in state INIT) (_DESTROY)
%7|1603848621.466|STATE|Tester#producer-1| [thrd::0/internal]: :0/internal: Broker changed state INIT -> DOWN
%7|1603848621.466|TERMINATE|Tester#producer-1| [thrd::0/internal]: :0/internal: Handle is terminating in state DOWN: 1 refcnts (0x563f68e17c80), 0 toppar(s), 0 active toppar(s), 0 outbufs, 0 waitresps, 0 retrybufs: failed 0 request(s) in retry+outbuf
%7|1603848621.467|TERM|Tester#producer-1| [thrd:kafka-2:29092/bootstrap]: kafka-2:29092/2: Received TERMINATE op in state UP: 1 refcnts, 0 toppar(s), 0 active toppar(s), 0 outbufs, 0 waitresps, 0 retrybufs
%7|1603848621.468|FAIL|Tester#producer-1| [thrd:kafka-2:29092/bootstrap]: kafka-2:29092/2: Client is terminating (after 6707ms in state UP) (_DESTROY)
%7|1603848621.468|STATE|Tester#producer-1| [thrd:kafka-2:29092/bootstrap]: kafka-2:29092/2: Broker changed state UP -> DOWN
%7|1603848621.468|FAIL|Tester#producer-1| [thrd::0/internal]: :0/internal: Broker handle is terminating (after 0ms in state DOWN) (_DESTROY)
%7|1603848621.469|TERMINATE|Tester#producer-1| [thrd:kafka-2:29092/bootstrap]: kafka-2:29092/2: Handle is terminating in state DOWN: 1 refcnts (0x563f68e19660), 0 toppar(s), 0 active toppar(s), 0 outbufs, 0 waitresps, 0 retrybufs: failed 0 request(s) in retry+outbuf
%7|1603848621.469|FAIL|Tester#producer-1| [thrd:kafka-2:29092/bootstrap]: kafka-2:29092/2: Broker handle is terminating (after 0ms in state DOWN) (_DESTROY)
```

Client logs when explicitly setting to None
=================================
```
Version: ('1.5.0', 17104896), Lib version: ('1.5.0', 17105151)
Inside tester() function. Admin client will be created here.
%7|1603848702.210|BRKMAIN|Tester#producer-1| [thrd::0/internal]: :0/internal: Enter main broker thread
%7|1603848702.210|BROKER|Tester#producer-1| [thrd:app]: kafka-1:19092/bootstrap: Added new broker with NodeId -1
%7|1603848702.211|BROKER|Tester#producer-1| [thrd:app]: kafka-2:29092/bootstrap: Added new broker with NodeId -1
%7|1603848702.211|BROKER|Tester#producer-1| [thrd:app]: kafka-3:39092/bootstrap: Added new broker with NodeId -1
%7|1603848702.211|CONNECT|Tester#producer-1| [thrd:app]: kafka-3:39092/bootstrap: Selected for cluster connection: bootstrap servers added (broker has 0 connection attempt(s))
%7|1603848702.211|BRKMAIN|Tester#producer-1| [thrd:kafka-1:19092/bootstrap]: kafka-1:19092/bootstrap: Enter main broker thread
%7|1603848702.212|BRKMAIN|Tester#producer-1| [thrd:kafka-2:29092/bootstrap]: kafka-2:29092/bootstrap: Enter main broker thread
%7|1603848702.212|BRKMAIN|Tester#producer-1| [thrd:kafka-3:39092/bootstrap]: kafka-3:39092/bootstrap: Enter main broker thread
%7|1603848702.212|CONNECT|Tester#producer-1| [thrd:kafka-3:39092/bootstrap]: kafka-3:39092/bootstrap: Received CONNECT op
%7|1603848702.212|STATE|Tester#producer-1| [thrd:kafka-3:39092/bootstrap]: kafka-3:39092/bootstrap: Broker changed state INIT -> TRY_CONNECT
%7|1603848702.212|INIT|Tester#producer-1| [thrd:app]: librdkafka v1.5.0 (0x10500ff) Tester#producer-1 initialized (builtin.features gzip,snappy,ssl,sasl,regex,lz4,sasl_plain,sasl_scram,plugins,zstd,sasl_oauthbearer, STATIC_LINKING GCC GXX PKGCONFIG INSTALL GNULD LDS LIBDL PLUGINS STATIC_LIB_zlib ZLIB STATIC_LIB_libcrypto STATIC_LIB_libssl SSL STATIC_LIB_libzstd ZSTD HDRHISTOGRAM SYSLOG SNAPPY SOCKEM SASL_SCRAM SASL_OAUTHBEARER CRC32C_HW, debug 0x2)
%7|1603848702.213|CONNECT|Tester#producer-1| [thrd:kafka-3:39092/bootstrap]: kafka-3:39092/bootstrap: broker in state TRY_CONNECT connecting
%7|1603848702.213|STATE|Tester#producer-1| [thrd:kafka-3:39092/bootstrap]: kafka-3:39092/bootstrap: Broker changed state TRY_CONNECT -> CONNECT
%7|1603848702.214|CONNECT|Tester#producer-1| [thrd:main]: Not selecting any broker for cluster connection: still suppressed for 47ms: lookup controller
%7|1603848702.215|CONNECT|Tester#producer-1| [thrd:kafka-3:39092/bootstrap]: kafka-3:39092/bootstrap: Connecting to ipv4#172.23.0.9:39092 (plaintext) with socket 11
%7|1603848702.215|CONNECT|Tester#producer-1| [thrd:kafka-3:39092/bootstrap]: kafka-3:39092/bootstrap: Connected to ipv4#172.23.0.9:39092
%7|1603848702.216|CONNECTED|Tester#producer-1| [thrd:kafka-3:39092/bootstrap]: kafka-3:39092/bootstrap: Connected (#1)
%7|1603848702.216|FEATURE|Tester#producer-1| [thrd:kafka-3:39092/bootstrap]: kafka-3:39092/bootstrap: Updated enabled protocol features +ApiVersion to ApiVersion
%7|1603848702.216|STATE|Tester#producer-1| [thrd:kafka-3:39092/bootstrap]: kafka-3:39092/bootstrap: Broker changed state CONNECT -> APIVERSION_QUERY
%7|1603848702.216|CONNECT|Tester#producer-1| [thrd:main]: Not selecting any broker for cluster connection: still suppressed for 44ms: lookup controller
%7|1603848702.222|FEATURE|Tester#producer-1| [thrd:kafka-3:39092/bootstrap]: kafka-3:39092/bootstrap: Updated enabled protocol features to MsgVer1,ApiVersion,BrokerBalancedConsumer,ThrottleTime,Sasl,SaslHandshake,BrokerGroupCoordinator,LZ4,OffsetTime,MsgVer2,IdempotentProducer,ZSTD,SaslAuthReq
%7|1603848702.222|STATE|Tester#producer-1| [thrd:kafka-3:39092/bootstrap]: kafka-3:39092/bootstrap: Broker changed state APIVERSION_QUERY -> UP
%7|1603848702.224|CLUSTERID|Tester#producer-1| [thrd:main]: kafka-3:39092/bootstrap: ClusterId update "" -> "JA1lgI6fTvWwm5lTvT7ETA"
%7|1603848702.224|CONTROLLERID|Tester#producer-1| [thrd:main]: kafka-3:39092/bootstrap: ControllerId update -1 -> 2
%7|1603848702.225|UPDATE|Tester#producer-1| [thrd:kafka-1:19092/bootstrap]: kafka-1:19092/bootstrap: NodeId changed from -1 to 1
%7|1603848702.225|UPDATE|Tester#producer-1| [thrd:kafka-1:19092/bootstrap]: kafka-1:19092/1: Name changed from kafka-1:19092/bootstrap to kafka-1:19092/1
%7|1603848702.225|LEADER|Tester#producer-1| [thrd:kafka-1:19092/bootstrap]: kafka-1:19092/1: Mapped 0 partition(s) to broker
%7|1603848702.226|UPDATE|Tester#producer-1| [thrd:kafka-2:29092/bootstrap]: kafka-2:29092/bootstrap: NodeId changed from -1 to 2
%7|1603848702.226|UPDATE|Tester#producer-1| [thrd:kafka-2:29092/bootstrap]: kafka-2:29092/2: Name changed from kafka-2:29092/bootstrap to kafka-2:29092/2
%7|1603848702.226|LEADER|Tester#producer-1| [thrd:kafka-2:29092/bootstrap]: kafka-2:29092/2: Mapped 0 partition(s) to broker
%7|1603848702.227|CONNECT|Tester#producer-1| [thrd:kafka-2:29092/bootstrap]: kafka-2:29092/2: Received CONNECT op
%7|1603848702.227|STATE|Tester#producer-1| [thrd:kafka-2:29092/bootstrap]: kafka-2:29092/2: Broker changed state INIT -> TRY_CONNECT
%7|1603848702.227|CONNECT|Tester#producer-1| [thrd:kafka-2:29092/bootstrap]: kafka-2:29092/2: broker in state TRY_CONNECT connecting
%7|1603848702.227|STATE|Tester#producer-1| [thrd:kafka-2:29092/bootstrap]: kafka-2:29092/2: Broker changed state TRY_CONNECT -> CONNECT
%7|1603848702.228|UPDATE|Tester#producer-1| [thrd:kafka-3:39092/bootstrap]: kafka-3:39092/bootstrap: NodeId changed from -1 to 3
%7|1603848702.228|UPDATE|Tester#producer-1| [thrd:kafka-3:39092/bootstrap]: kafka-3:39092/3: Name changed from kafka-3:39092/bootstrap to kafka-3:39092/3
%7|1603848702.228|LEADER|Tester#producer-1| [thrd:kafka-3:39092/bootstrap]: kafka-3:39092/3: Mapped 0 partition(s) to broker
%7|1603848702.228|STATE|Tester#producer-1| [thrd:kafka-3:39092/bootstrap]: kafka-3:39092/3: Broker changed state UP -> UPDATE
%7|1603848702.228|CONNECT|Tester#producer-1| [thrd:kafka-2:29092/bootstrap]: kafka-2:29092/2: Connecting to ipv4#172.23.0.7:29092 (plaintext) with socket 12
%7|1603848702.229|CONNECT|Tester#producer-1| [thrd:kafka-2:29092/bootstrap]: kafka-2:29092/2: Connected to ipv4#172.23.0.7:29092
%7|1603848702.229|CONNECTED|Tester#producer-1| [thrd:kafka-2:29092/bootstrap]: kafka-2:29092/2: Connected (#1)
%7|1603848702.229|FEATURE|Tester#producer-1| [thrd:kafka-2:29092/bootstrap]: kafka-2:29092/2: Updated enabled protocol features +ApiVersion to ApiVersion
%7|1603848702.230|STATE|Tester#producer-1| [thrd:kafka-2:29092/bootstrap]: kafka-2:29092/2: Broker changed state CONNECT -> APIVERSION_QUERY
%7|1603848702.231|STATE|Tester#producer-1| [thrd:kafka-3:39092/bootstrap]: kafka-3:39092/3: Broker changed state UPDATE -> UP
%7|1603848702.233|FEATURE|Tester#producer-1| [thrd:kafka-2:29092/bootstrap]: kafka-2:29092/2: Updated enabled protocol features to MsgVer1,ApiVersion,BrokerBalancedConsumer,ThrottleTime,Sasl,SaslHandshake,BrokerGroupCoordinator,LZ4,OffsetTime,MsgVer2,IdempotentProducer,ZSTD,SaslAuthReq
%7|1603848702.233|STATE|Tester#producer-1| [thrd:kafka-2:29092/bootstrap]: kafka-2:29092/2: Broker changed state APIVERSION_QUERY -> UP
Topic already exists.
Manually setting admin client to None
%7|1603848703.242|DESTROY|Tester#producer-1| [thrd:app]: Terminating instance (destroy flags none (0x0))
%7|1603848703.244|DESTROY|Tester#producer-1| [thrd:main]: Destroy internal
%7|1603848703.246|DESTROY|Tester#producer-1| [thrd:main]: Waiting for background queue thread to terminate
%7|1603848703.246|DESTROY|Tester#producer-1| [thrd:main]: Removing all topics
%7|1603848703.246|DESTROY|Tester#producer-1| [thrd:main]: Sending TERMINATE to kafka-3:39092/3
%7|1603848703.246|DESTROY|Tester#producer-1| [thrd:main]: Sending TERMINATE to kafka-2:29092/2
%7|1603848703.247|DESTROY|Tester#producer-1| [thrd:main]: Sending TERMINATE to kafka-1:19092/1
%7|1603848703.247|TERM|Tester#producer-1| [thrd:kafka-1:19092/bootstrap]: kafka-1:19092/1: Received TERMINATE op in state INIT: 1 refcnts, 0 toppar(s), 0 active toppar(s), 0 outbufs, 0 waitresps, 0 retrybufs
%7|1603848703.247|FAIL|Tester#producer-1| [thrd:kafka-1:19092/bootstrap]: kafka-1:19092/1: Client is terminating (after 1036ms in state INIT) (_DESTROY)
%7|1603848703.248|STATE|Tester#producer-1| [thrd:kafka-1:19092/bootstrap]: kafka-1:19092/1: Broker changed state INIT -> DOWN
%7|1603848703.248|TERMINATE|Tester#producer-1| [thrd:kafka-1:19092/bootstrap]: kafka-1:19092/1: Handle is terminating in state DOWN: 1 refcnts (0x55623d58d970), 0 toppar(s), 0 active toppar(s), 0 outbufs, 0 waitresps, 0 retrybufs: failed 0 request(s) in retry+outbuf
%7|1603848703.250|TERM|Tester#producer-1| [thrd:kafka-2:29092/bootstrap]: kafka-2:29092/2: Received TERMINATE op in state UP: 1 refcnts, 0 toppar(s), 0 active toppar(s), 0 outbufs, 0 waitresps, 0 retrybufs
%7|1603848703.253|FAIL|Tester#producer-1| [thrd:kafka-2:29092/bootstrap]: kafka-2:29092/2: Client is terminating (after 1019ms in state UP) (_DESTROY)
%7|1603848703.254|STATE|Tester#producer-1| [thrd:kafka-2:29092/bootstrap]: kafka-2:29092/2: Broker changed state UP -> DOWN
%7|1603848703.256|TERMINATE|Tester#producer-1| [thrd:kafka-2:29092/bootstrap]: kafka-2:29092/2: Handle is terminating in state DOWN: 1 refcnts (0x55623d58e660), 0 toppar(s), 0 active toppar(s), 0 outbufs, 0 waitresps, 0 retrybufs: failed 0 request(s) in retry+outbuf
%7|1603848703.258|FAIL|Tester#producer-1| [thrd:kafka-2:29092/bootstrap]: kafka-2:29092/2: Broker handle is terminating (after 2ms in state DOWN) (_DESTROY)
%7|1603848703.251|TERM|Tester#producer-1| [thrd::0/internal]: :0/internal: Received TERMINATE op in state INIT: 1 refcnts, 0 toppar(s), 0 active toppar(s), 0 outbufs, 0 waitresps, 0 retrybufs
%7|1603848703.250|FAIL|Tester#producer-1| [thrd:kafka-1:19092/bootstrap]: kafka-1:19092/1: Broker handle is terminating (after 0ms in state DOWN) (_DESTROY)
%7|1603848703.251|TERM|Tester#producer-1| [thrd:kafka-3:39092/bootstrap]: kafka-3:39092/3: Received TERMINATE op in state UP: 1 refcnts, 0 toppar(s), 0 active toppar(s), 0 outbufs, 0 waitresps, 0 retrybufs
%7|1603848703.262|FAIL|Tester#producer-1| [thrd:kafka-3:39092/bootstrap]: kafka-3:39092/3: Client is terminating (after 1031ms in state UP) (_DESTROY)
%7|1603848703.262|FAIL|Tester#producer-1| [thrd::0/internal]: :0/internal: Client is terminating (after 1051ms in state INIT) (_DESTROY)
%7|1603848703.263|STATE|Tester#producer-1| [thrd:kafka-3:39092/bootstrap]: kafka-3:39092/3: Broker changed state UP -> DOWN
%7|1603848703.266|TERMINATE|Tester#producer-1| [thrd:kafka-3:39092/bootstrap]: kafka-3:39092/3: Handle is terminating in state DOWN: 1 refcnts (0x55623d58f410), 0 toppar(s), 0 active toppar(s), 0 outbufs, 0 waitresps, 0 retrybufs: failed 0 request(s) in retry+outbuf
%7|1603848703.266|FAIL|Tester#producer-1| [thrd:kafka-3:39092/bootstrap]: kafka-3:39092/3: Broker handle is terminating (after 0ms in state DOWN) (_DESTROY)
%7|1603848703.264|STATE|Tester#producer-1| [thrd::0/internal]: :0/internal: Broker changed state INIT -> DOWN
%7|1603848703.267|TERMINATE|Tester#producer-1| [thrd::0/internal]: :0/internal: Handle is terminating in state DOWN: 1 refcnts (0x55623d58cc80), 0 toppar(s), 0 active toppar(s), 0 outbufs, 0 waitresps, 0 retrybufs: failed 0 request(s) in retry+outbuf
%7|1603848703.268|FAIL|Tester#producer-1| [thrd::0/internal]: :0/internal: Broker handle is terminating (after 1ms in state DOWN) (_DESTROY)
Last line of tester() function. Admin client should go out of scope here.
Tester function has returned. All variables should be GC'ed by now.
AdminClient in GC-0 False
AdminClient in GC-1 False
AdminClient in GC-2 False
^CDone
```

Checklist
=========
Please provide the following information:

 - [ ] confluent-kafka-python and librdkafka version (`('1.5.0', 17104896)` and `('1.5.0', 17105151)`):
 - [ ] Apache Kafka broker version: 5.5.1
 - [ ] Client configuration: `{"bootstrap.servers": "kafka-1:19092,kafka-2:29092,kafka-3:39092", "client.id": "Tester", "debug": "broker"}`
 - [ ] Operating system: GNU/Linux x86_64 (Kernel version 4.19.76)
 - [ ] Provide broker log excerpts: N/A (seems like a client issue)
 - [ ] Critical issue (no, workaround seems to exist)

**Comment by ckataki** [comment]
We did some more digging, and this problem seems to be with `concurrent.futures.Future` instead. We were able to reproduce this behavior without the Confluent libraries (see https://bugs.python.org/issue35930).
Replacing the try-catch block around the future:
```
try:
    future.result()
except Exception as exc:
    print(exc)
```
with something like this:
```
if future.exception():
    print(future.exception())
else:
    print(future.result())
```
This seems to solve our issue. We will try it out overnight, and close this issue tomorrow.
